# CISC-873-DM-W23-A2 ❤️
## This Notebook is divided into two parts:
#### 1. Answring some questions about data mining subject. 🤓
#### 2. A2 Speed Dating Match Prediction problem. 
##### Kaggle account name: Manar Elghobashy.

## 1. Answring some questions about data mining subject🤓:
-----------------------------------------------------------------------
#### Q1:Why a simple linear regression model (without any activation function) is not good for classification task, compared to Perceptron/Logistic regression?
* Regression problems are solved using linear regression, while classification problems are solved using logistic regression.
* Linear regression provides a continuous output but Logistic regression provides discreet output.
* Linear regression's goal is to determine the best-fitting line, but logistic regression goes one step farther and fits the line values to the sigmoid curve.
* The method for calculating loss function in linear regression is the mean squared error whereas for logistic regression it is maximum likelihood estimation.
* When new data points are added, the threshold value shifts in linear regression.
-----------------------------------------------------------------------
#### Q2: What's a decision tree and how it is different to a logistic regression model?
* A decision tree is a flowchart-like tree structure in which each internal node represents an attribute test, each branch reflects the test's result, and each leaf node (terminal node) stores a class label.For categorization and prediction, the decision tree is the most powerful and widely used technique.
and consists of

  * Root node: The base of the decision tree.

  * Splitting: The process of dividing a node into multiple sub-nodes.

  * Branch: A subsection of the decision tree consisting of multiple nodes.

  * Decision node: When a sub-node is further split into additional sub-nodes.

  * Leaf node: When a sub-node does not further split into additional sub-nodes; represents possible outcomes.
* difference between it and logistic regression model:


> * The Decision Tree is easier to understand. However, Logistic Regression is more difficult to interpret.
* The Decision Tree's Decision Boundaries are The space is divided into smaller areas by bisection. However, Logistic Regression Decision Boundaries are Linear and have only one decision boundary.
* The Decision Tree's Decision Making Ease is Decision-making is handled automatically. However, Logistic Regression Decision Making Ease is A decision threshold must be established.
* Decision Tree is Prone to overfitting. but Logistic Regression isn't prone to overfitting.
* Noise has a significant impact on the Decision Tree's robustness to noise. However, Logistic Regression's robustness to noise is robust.
* Decision Tree Scalability It's possible to train it on a modest training set. Scalability of Logistic Regression, on the other hand, necessitates a large enough training set.
---------------------------------------------------------------------------
#### Q3: What's the difference between grid search and random search?
* both are mainly used to tune hyperparameters of any model, but the difference between both, is that the grid search for the best combination from all hyperparameter combinations, every combination of hyperparameter values is tried which can be very inefficient and it will take a long time to execute, in other hand random search sets up a grid of hyperparameter values and searches for random combinations to train the model and score, it will not find the best hyperparameter as grid search do, but it's faster than grid search.
* **Grid search** 
  * Try out every combination of the parameters
  * Computationally expensive
  * Global optimal (within the given range)
  * Sklearn: model_selection.GridSearchCV
* **Random search**
  * Try out a random subset
  * good enough
  * Local optimal (within the given range)
  * Efficient (less trials)
  * Sklearn: model_selection.RandomizedSearchCV
-----------------------------------------------------------------------------
#### Q4: What's the difference between bayesian search and random search?
* Bayesian search is a sequential model-based optimization algorithm that uses the results from the previous iteration to decide the next hyperparameter value while randoms search sets up a grid of hyperparameter values and selects random combinations to train the model and score (provide a statistical distribution for each hyperparameter from which values are sampled) bayesian search is performing better than random search in finding best hyperparameters and it's faster than grid search.
* **Random Search**
  * Try out a random subset
  * good enough
  * Local optimal (within the given range)
  * random search are completely uninformed by past evaluations.
  * Efficient (less trials)
  * Sklearn: model_selection.RandomizedSearchCV
* **Bayesian Optimization**
  * As an optimization problem
  * it use to pick the next set of hyperparameters which will improve the model performance. We iteratively repeat this process until we converge to an optimum.(Trial -> estimated error -> Bayesian model estimates the next parameter to try -> trial -> repeat..)

  * Bayesian methods can find the best hyperparameters in lesser time (in fewer iterations) than random search.
  *pip install bayesian-optimization
--------------------------------------------------------------------------------



## 2. A2 Speed Dating Match Prediction problem.
## Problem Formulation
### Define the problem
* This notebook will look into if dating preferences, attribute ratings, and background information can be used to forecast whether a partner will match with their date.
* The problem can be framed as a supervised, binary classification problem where the model predicts if a partner has accepted or rejected their date.

### What is the input?
* All Dataset without the match columns (191 number of input features).

### The Dataset:

* Participants attended a dating event where they had a 4-minute date with every other participant of the opposite sex who attended the same event.
The participants decided to accept or reject their partner. If both the participant and partner matched, they received each other's contact information.
Participants rated their partners on six personal attributes: attractiveness, sincerity, intelligence, fun, ambition and shared interests.
Before and after the event, participants rated their preferences in the six attributes and gave themselves ratings.
* Other information was collected about the participants' background and preferences.

### What is the output?
* The participants decided to accept or reject their partner. If both the participant and partner matched. In the dataset, the match column is output.

### What data mining function is required?
* In this case, it will be binary Classification that separates data points into different classes (accept or reject / 1 or 0) which If both the participant and partner matched.

### What could be the challenges?
* the datasets have multiplue missing values.
* the datasets are highly unbalanced.
* predict the match betweem the participant and partner is correct.

### What is the impact?
* creating a new system and give it some Features, it can decide whether or not people are compatible with one another. without these individuals coming together.

### What is an ideal solution?
According to my subsequent attempts, Bayesian Search and XGBoost Classifier with Cross Validation is the best approach because it provides me the highest kaggle score.

* In Kaggle 
  * Public score: 0.88296
  * Private score: 0.89129


### In this notebook:
#### 1. data preprocessing
#### 2. pipeline and models


> * XGB model pipeline
* Random forest
* Logistic Regression



### 1. data preprocessing

1.1 importing required libraries.

In [ ]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O.
import matplotlib.pyplot as plt # visualization library. 
import seaborn as sns
# some important libraries that is requried for building pipeline and models.
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.datasets import fetch_openml
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split, GridSearchCV, RandomizedSearchCV
from xgboost.sklearn import XGBClassifier
pd.set_option("display.max_rows", None)
pd.set_option("display.max_columns", None)

1.2 reading the train/test csv files using pandas 


*   read the files of training and testing data
*   save the id column before dropping it as we needed in the submission file.



In [ ]:
#reading train and test data csv files.
train_data=pd.read_csv('train.csv')
test_data=pd.read_csv('test.csv')
# save the id column before dropping it.
train_id=train_data['id']
test_id=test_data['id']
#drop the id column fro train and test.
train_data=train_data.drop(['id'],axis=1)
test_data=test_data.drop(['id'],axis=1)

save the match column which is the label column in variable y and dropped to get only the features in the train_data variable.

In [ ]:
#split the train data into train_data containing features and labels.
y=train_data['match']
train_data=train_data.drop(['match'],axis=1)

concatenate both train and test to do the preprocessing on both at atime

In [ ]:
# train and test concatination to do all preprocessing on them at once.
all_data = pd.concat([train_data,test_data],axis = 0)
print(all_data.shape[0], 'observations and', all_data.shape[1], 'features')
all_data.shape

8378 observations and 190 features


(8378, 190)

In [ ]:
# display some of the data to see the features and their values.
all_data.head(5)

gender  idg  condtn  wave  round  position  positin1  order  partner  \
0       0    3       2    14     18         2       2.0     14       12   
1       1   14       1     3     10         2       NaN      8        8   
2       1   14       1    13     10         8       8.0     10       10   
3       1   38       2     9     20        18      13.0      6        7   
4       1   24       2    14     20         6       6.0     20       17   

     pid  int_corr  samerace  age_o  race_o  pf_o_att  pf_o_sin  pf_o_int  \
0  372.0     -0.03         0   27.0     2.0     30.00     15.00     15.00   
1   63.0      0.21         0   24.0     4.0      5.00     15.00     45.00   
2  331.0      0.43         0   34.0     2.0     15.00     15.00     10.00   
3  200.0      0.72         1   25.0     2.0     13.21     18.87     18.87   
4  357.0      0.33         0   27.0     4.0     15.00     20.00     20.00   

   pf_o_fun  pf_o_amb  pf_o_sha  attr_o  sinc_o  intel_o  fun_o  amb_o  \
0     20.00      5.00     15.00     7.0     7.0      7.0    6.0    5.0   
1     25.00      0.00     10.00     3.0     8.0      5.0    3.0    7.0   
2     25.00     10.00     25.00     4.0     8.0      7.0    4.0    7.0   
3     16.98     16.98     15.09     5.0     9.0      7.0    5.0    8.0   
4     20.00     20.00      5.00     4.0     5.0      7.0    5.0    5.0   

   shar_o  like_o  prob_o  met_o   age  \
0     NaN     7.0     1.0    2.0  33.0   
1     1.0     1.0     3.0    2.0  22.0   
2     3.0     3.0     2.0    2.0  27.0   
3     NaN     7.0     6.0    2.0  31.0   
4     6.0     4.0     3.0    2.0  27.0   

                                    field  field_cd  \
0  Ed.D. in higher education policy at TC       9.0   
1                             Engineering       5.0   
2                          Urban Planning       5.0   
3                   International Affairs      13.0   
4                                Business       8.0   

                                           undergra    mn_sat    tuition  \
0                  University of Michigan-Ann Arbor  1,290.00  21,645.00   
1                                               NaN       NaN        NaN   
2  Rizvi College of Architecture, Bombay University       NaN        NaN   
3                                               NaN       NaN        NaN   
4                                   Harvard College  1,400.00  26,019.00   

   race  imprace  imprelig            from zipcode     income  goal  date  \
0   3.0      2.0       1.0   Palo Alto, CA     NaN        NaN   1.0   6.0   
1   2.0      8.0       1.0      Boston, MA   2,021        NaN   5.0   6.0   
2   6.0      1.0       1.0   Bombay, India     NaN        NaN   1.0   4.0   
3   2.0      4.0       7.0  Washington, DC  10,471  45,300.00   2.0   5.0   
4   2.0      9.0       7.0     Midwest USA  66,208  46,138.00   2.0   5.0   

   go_out                             career  career_c  sports  tvsports  \
0     3.0               University President       2.0     3.0       4.0   
1     1.0  Engineer or iBanker or consultant       7.0     8.0       3.0   
2     2.0             Real Estate Consulting       7.0     4.0       2.0   
3     4.0                     public service       9.0     5.0       5.0   
4     2.0                          undecided      10.0     9.0      10.0   

   exercise  dining  museums   art  hiking  gaming  clubbing  reading    tv  \
0       4.0    10.0      9.0  10.0     1.0     1.0       8.0      7.0  10.0   
1       7.0    10.0      6.0   6.0     7.0     2.0       9.0      5.0   6.0   
2       7.0     6.0      8.0   8.0     8.0     6.0       6.0      8.0   4.0   
3       9.0     8.0      7.0   8.0     9.0     3.0       8.0      7.0   4.0   
4       8.0     9.0      9.0   9.0     7.0     4.0       2.0      9.0   6.0   

   theater  movies  concerts  music  shopping  yoga  exphappy  expnum  \
0      8.0    10.0       1.0    4.0      10.0   4.0       2.0     NaN   
1      7.0     7.0       6.0    8.0       8.0   4.0       7.0 

In [ ]:
# display information about the training  data 
train_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5909 entries, 0 to 5908
Columns: 190 entries, gender to amb5_3
dtypes: float64(173), int64(9), object(8)
memory usage: 8.6+ MB


In [ ]:
# see the shape of the train and test data so that i can split the merged data after the preprocessing.
print(train_data.shape)
print(test_data.shape)


(5909, 190)
(2469, 190)


check null values and drop columns with 50% null values

In [ ]:
# see if the features have null values and display them sorted. 
all_data.isnull().sum().sort_values()

gender         0
partner        0
order          0
position       0
samerace       0
wave           0
condtn         0
idg            0
round          0
pid           10
race          63
field         63
race_o        73
art           79
go_out        79
sports        79
attr2_1       79
tvsports      79
exercise      79
dining        79
museums       79
goal          79
hiking        79
movies        79
clubbing      79
reading       79
tv            79
theater       79
sinc2_1       79
concerts      79
music         79
shopping      79
yoga          79
attr1_1       79
sinc1_1       79
intel1_1      79
gaming        79
from          79
intel2_1      79
imprace       79
fun2_1        79
imprelig      79
field_cd      82
fun1_1        89
amb2_1        89
career        89
pf_o_att      89
shar2_1       89
pf_o_sin      89
pf_o_int      89
age           95
date          97
pf_o_fun      98
amb1_1        99
exphappy     101
age_o        104
fun3_1       105
sinc3_1      105
attr3_1      1

In [ ]:
# apply a threshold to drop the columns with null values up to 50%.
threshold = len(all_data) * .5
all_data = all_data.dropna(thresh = threshold, axis = 1)
# check the rest of the features.
all_data.isnull().sum().sort_values()


gender         0
samerace       0
order          0
position       0
partner        0
wave           0
condtn         0
idg            0
round          0
pid           10
field         63
race          63
race_o        73
gaming        79
imprace       79
hiking        79
imprelig      79
dining        79
art           79
museums       79
goal          79
clubbing      79
go_out        79
exercise      79
from          79
reading       79
music         79
theater       79
fun2_1        79
intel2_1      79
sinc2_1       79
attr2_1       79
intel1_1      79
tv            79
sinc1_1       79
yoga          79
shopping      79
tvsports      79
concerts      79
movies        79
attr1_1       79
sports        79
field_cd      82
amb2_1        89
shar2_1       89
fun1_1        89
career        89
pf_o_att      89
pf_o_sin      89
pf_o_int      89
age           95
date          97
pf_o_fun      98
amb1_1        99
exphappy     101
age_o        104
amb3_1       105
intel3_1     105
sinc3_1      1

In [ ]:
# display the shape of the data again to see how many columns are lefted after dropping the null columns.
all_data.shape

(8378, 131)

check duplications in the data

In [ ]:
# check if there are any duplication.
all_data.duplicated().sum()

0

In [ ]:
# see the type of features to work on the properate preprcessing for them.
all_data.dtypes.value_counts()

float64    116
int64        9
object       6
dtype: int64

In [ ]:
# show all object columns that need to be converted by one hot encoder.
all_data[all_data.select_dtypes(['object']).columns].columns

Index(['field', 'undergra', 'from', 'zipcode', 'income', 'career'], dtype='object')

display correlation map and drop uncorrelated columns

In [ ]:
# see the correlation map to detect the correlated features and drop uncorrelated ones.
corr = all_data.corr()

corr.style.background_gradient(cmap="BuPu")

In [ ]:
# drop some feature as they aren't correlated.
all_data = all_data.drop(['idg', 'pid', 'zipcode'], axis=1)

convert objects column to categorical ones.

In [ ]:
#display unique values of the object feature 'field'
all_data['field'].unique()

array(['Ed.D. in higher education policy at TC', 'Engineering',
       'Urban Planning', 'International Affairs', 'Business', 'MBA',
       'Operations Research (SEAS)', 'french', 'nutrition',
       'MA Teaching Social Studies', 'Education',
       'Electrical Engineering', 'Operations Research',
       'electrical engineering', 'business', 'sociology', 'Law',
       'Ecology', 'Genetics', 'School Psychology', 'Human Rights',
       'MA in Quantitative Methods', 'Mathematics', 'Climate Dynamics',
       'Finance/Economics', 'LAW', 'History',
       'Biochemistry & Molecular Biophysics', 'Social Work',
       'MA Science Education', 'film', 'law',
       'International Security Policy - SIPA', 'Sociology', 'Physics',
       'medicine', 'International Development', 'climate change',
       'Curriculum and Teaching/Giftedness',
       'Master in Public Administration', 'Neurobiology', 'Biology',
       'Biomedical Engineering', 'microbiology', 'Mathematics, PhD',
       'History of Relig

In [ ]:
# see how many are these unique values.
all_data['field'].nunique()

259

In [ ]:
# assign most of the object features to categorical one.
all_data['field'] = all_data['field'].astype('category')

all_data['undergra']= all_data['undergra'].astype('category')

all_data['from'] = all_data['from'].astype('category')

all_data['career'] = all_data['career'].astype('category')


In [ ]:
# display features type again.
all_data.dtypes.value_counts()

float64     115
int64         8
category      1
category      1
category      1
object        1
category      1
dtype: int64

covert income column into type float

In [ ]:
all_data['income']

0              NaN
1              NaN
2              NaN
3        45,300.00
4        46,138.00
5              NaN
6        55,080.00
7              NaN
8              NaN
9        36,594.00
10       39,394.00
11       16,908.00
12             NaN
13             NaN
14             NaN
15             NaN
16             NaN
17             NaN
18       49,642.00
19       23,361.00
20             NaN
21       46,800.00
22             NaN
23       40,749.00
24       48,944.00
25       63,351.00
26       75,347.00
27             NaN
28             NaN
29             NaN
30       42,640.00
31             NaN
32       48,944.00
33       57,513.00
34       39,919.00
35       32,649.00
36       32,508.00
37      109,031.00
38       37,881.00
39             NaN
40       37,754.00
41             NaN
42       28,989.00
43       55,080.00
44       28,759.00
45             NaN
46             NaN
47             NaN
48       55,080.00
49       32,649.00
50             NaN
51             NaN
52          

In [ ]:
#converting the object column 'income' to a float feature by removing ',' at first then assign it to type float 
all_data.income = all_data.income.str.replace(',', '')
all_data['income'] = all_data['income'].astype('float64')
all_data['income']

0            NaN
1            NaN
2            NaN
3        45300.0
4        46138.0
5            NaN
6        55080.0
7            NaN
8            NaN
9        36594.0
10       39394.0
11       16908.0
12           NaN
13           NaN
14           NaN
15           NaN
16           NaN
17           NaN
18       49642.0
19       23361.0
20           NaN
21       46800.0
22           NaN
23       40749.0
24       48944.0
25       63351.0
26       75347.0
27           NaN
28           NaN
29           NaN
30       42640.0
31           NaN
32       48944.0
33       57513.0
34       39919.0
35       32649.0
36       32508.0
37      109031.0
38       37881.0
39           NaN
40       37754.0
41           NaN
42       28989.0
43       55080.0
44       28759.0
45           NaN
46           NaN
47           NaN
48       55080.0
49       32649.0
50           NaN
51           NaN
52           NaN
53       46800.0
54           NaN
55           NaN
56       55080.0
57       45695.0
58       64716

### In part 1 "the preprocessing" 
* multiple functions are applied on the data as:
  * display null values in each column.
  * check if there are any duplications.
  * display the correlation map.
  * drop columns with null values above 50% and uncorrelated columns.
  * set objects columns as categorical ones.
  * covert income column from object to float.

### 2. pipeline and models
this pert will be divided into preparing data before using it in the models by spliting it the building the pipelines and apply three different model on this data:
* XGB model two trials: without oversampling, with oversampling
* Random forest
* Logistic Regression 

1. split data

In [ ]:
# splitting the merged data to train and test again.
train_data = all_data.iloc[:train_data.shape[0],:]
test_data = all_data.iloc[train_data.shape[0]:,:]
train_data.head()

gender  condtn  wave  round  position  positin1  order  partner  int_corr  \
0       0       2    14     18         2       2.0     14       12     -0.03   
1       1       1     3     10         2       NaN      8        8      0.21   
2       1       1    13     10         8       8.0     10       10      0.43   
3       1       2     9     20        18      13.0      6        7      0.72   
4       1       2    14     20         6       6.0     20       17      0.33   

   samerace  age_o  race_o  pf_o_att  pf_o_sin  pf_o_int  pf_o_fun  pf_o_amb  \
0         0   27.0     2.0     30.00     15.00     15.00     20.00      5.00   
1         0   24.0     4.0      5.00     15.00     45.00     25.00      0.00   
2         0   34.0     2.0     15.00     15.00     10.00     25.00     10.00   
3         1   25.0     2.0     13.21     18.87     18.87     16.98     16.98   
4         0   27.0     4.0     15.00     20.00     20.00     20.00     20.00   

   pf_o_sha  attr_o  sinc_o  intel_o  fun_o  amb_o  shar_o  like_o  prob_o  \
0     15.00     7.0     7.0      7.0    6.0    5.0     NaN     7.0     1.0   
1     10.00     3.0     8.0      5.0    3.0    7.0     1.0     1.0     3.0   
2     25.00     4.0     8.0      7.0    4.0    7.0     3.0     3.0     2.0   
3     15.09     5.0     9.0      7.0    5.0    8.0     NaN     7.0     6.0   
4      5.00     4.0     5.0      7.0    5.0    5.0     6.0     4.0     3.0   

   met_o   age                                   field  field_cd  \
0    2.0  33.0  Ed.D. in higher education policy at TC       9.0   
1    2.0  22.0                             Engineering       5.0   
2    2.0  27.0                          Urban Planning       5.0   
3    2.0  31.0                   International Affairs      13.0   
4    2.0  27.0                                Business       8.0   

                                           undergra  race  imprace  imprelig  \
0                  University of Michigan-Ann Arbor   3.0      2.0       1.0   
1                                               NaN   2.0      8.0       1.0   
2  Rizvi College of Architecture, Bombay University   6.0      1.0       1.0   
3                                               NaN   2.0      4.0       7.0   
4                                   Harvard College   2.0      9.0       7.0   

             from   income  goal  date  go_out  \
0   Palo Alto, CA      NaN   1.0   6.0     3.0   
1      Boston, MA      NaN   5.0   6.0     1.0   
2   Bombay, India      NaN   1.0   4.0     2.0   
3  Washington, DC  45300.0   2.0   5.0     4.0   
4     Midwest USA  46138.0   2.0   5.0     2.0   

                              career  career_c  sports  tvsports  exercise  \
0               University President       2.0     3.0       4.0       4.0   
1  Engineer or iBanker or consultant       7.0     8.0       3.0       7.0   
2             Real Estate Consulting       7.0     4.0       2.0       7.0   
3                     public service       9.0     5.0       5.0       9.0   
4                          undecided      10.0     9.0      10.0       8.0   

   dining  museums   art  hiking  gaming  clubbing  reading    tv  theater  \
0    10.0      9.0  10.0     1.0     1.0       8.0      7.0  10.0      8.0   
1    10.0      6.0   6.0     7.0     2.0       9.0      5.0   6.0      7.0   
2     6.0      8.0   8.0     8.0     6.0       6.0      8.0   4.0     10.0   
3     8.0      7.0   8.0     9.0     3.0       8.0      7.0   4.0      6.0   
4     9.0      9.0   9.0     7.0     4.0       2.0      9.0   6.0      8.0   

   movies  concerts  music  shopping  yoga  exphappy  attr1_1  sinc1_1  \
0    10.0       1.0    4.0      10.0   4.0       2.0    40.00    15.00   
1     7.0       6.0    8.0       8.0   4.0       7.0    21.00    20.00   
2    10.0       7.0    7.0       5.0  10.0       2.0    10.00    20.00   
3     8.0       7.0    7.0       3.0   5.0       4.0    17.78    17.78   
4     6.0       7.0    7.0       4.0   2.0       5.0    30.00    15.00   

In [ ]:
train_data.shape

(5909, 128)

2. build the pipelines 

In [ ]:
# we extract numeric features and categorical features names
# for later use

# numeric features can be selected by: (based on the df2.info() output )
features_numeric = list(train_data.select_dtypes(include=['float64', 'int64']))

# categorical features can be selected by: (based on the df2.info() output )
features_categorical = list(train_data.select_dtypes(include=['category']))

print('numeric features:', features_numeric)
print('categorical features:', features_categorical)

numeric features: ['gender', 'condtn', 'wave', 'round', 'position', 'positin1', 'order', 'partner', 'int_corr', 'samerace', 'age_o', 'race_o', 'pf_o_att', 'pf_o_sin', 'pf_o_int', 'pf_o_fun', 'pf_o_amb', 'pf_o_sha', 'attr_o', 'sinc_o', 'intel_o', 'fun_o', 'amb_o', 'shar_o', 'like_o', 'prob_o', 'met_o', 'age', 'field_cd', 'race', 'imprace', 'imprelig', 'income', 'goal', 'date', 'go_out', 'career_c', 'sports', 'tvsports', 'exercise', 'dining', 'museums', 'art', 'hiking', 'gaming', 'clubbing', 'reading', 'tv', 'theater', 'movies', 'concerts', 'music', 'shopping', 'yoga', 'exphappy', 'attr1_1', 'sinc1_1', 'intel1_1', 'fun1_1', 'amb1_1', 'shar1_1', 'attr4_1', 'sinc4_1', 'intel4_1', 'fun4_1', 'amb4_1', 'shar4_1', 'attr2_1', 'sinc2_1', 'intel2_1', 'fun2_1', 'amb2_1', 'shar2_1', 'attr3_1', 'sinc3_1', 'fun3_1', 'intel3_1', 'amb3_1', 'attr5_1', 'sinc5_1', 'intel5_1', 'fun5_1', 'amb5_1', 'attr', 'sinc', 'intel', 'fun', 'amb', 'shar', 'like', 'prob', 'met', 'match_es', 'satis_2', 'length', 'numdat_

in this part a multiple pipelines are built to preprocess the data by:
* for numeric data --> a pipeline called transformer_numeric will fill null data and apply standard scaler on the data.
* for categorical data --> a pipeline called transformer_categorical will fill the null using simple imputer the will apply one hot encoder on the categorical data.
* a pipieline called column_transformer will call the above pipeline to preprocess the data

after that, pipline will be created and set multiple classifiers and fit them, predict the testing set and record the probability of prediction in csv file.

In [ ]:
np.random.seed(0)

# define a pipe line for numeric feature preprocessing
# we gave them a name so we can set their hyperparameters
# it will use simple imputer to fill null values in the data.
transformer_numeric = Pipeline(
    steps=[
        ('imputer', SimpleImputer()),
        ('scaler', StandardScaler())]
)

# define a pipe line for categorical feature preprocessing
# we gave them a name so we can set their hyperparameters
transformer_categorical = Pipeline(
    steps=[
        ('imputer', SimpleImputer(strategy='constant')),
        ('onehot', OneHotEncoder(handle_unknown='ignore'))
    ]
)
# define the preprocessor 
# we gave them a name so we can set their hyperparameters
# we also specify what are the categorical 
preprocessor = ColumnTransformer(
    transformers=[
        ('num', transformer_numeric, features_numeric),
        ('cat', transformer_categorical, features_categorical)
    ]
)


#### Trial 1 Xgboost pipeline
This algorithm goes by lots of different names such as gradient boosting, multiple additive regression trees, stochastic gradient boosting or gradient boosting machines.

A Gradient Boosting Decision Trees (GBDT) is a decision tree ensemble learning algorithm similar to random forest, Ensemble learning algorithms combine multiple machine learning algorithms to obtain a better model. Random forest uses to build full decision trees in parallel from random bootstrap samples of the data set.

It is the top machine learning library for regression, classification, and ranking tasks.

* It includes parallel tree boosting.
* It supports regularization.
* It is designed to handle missing data with its in-build features.
* The user can run a cross-validation after each iteration.
* It works well in small to medium dataset.
* It is designed to be highly efficient, flexible and portable.
* It has a distributed weighted quantile sketch algorithm to effectively handle weighted data.
* To develop a model, the XGBoost classifier contains a lot of hyperparameters. 

I'll use some of them to assist us enhance the model and score.

In [ ]:
# building the XGB model pipeline.
# it will use the preprocessor pipeline so that the data will be preprocessed.
XGB_pipeline = Pipeline(
    steps=[
        ('preprocessor', preprocessor),
        ('my_classifier', 
           XGBClassifier(class_weight="balanced"),
        )
    ]
)
XGB_pipeline

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer()),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  ['gender', 'condtn', 'wave',
                                                   'round', 'position',
                                                   'positin1', 'order',
                                                   'partner', 'int_corr',
                                                   'samerace', 'age_o',
                                                   'race_o', 'pf_o_att',
                                                   'pf_o_sin', 'pf_o_int',
                                                   'pf_o_fun', 'pf_o_amb',
                                                   'pf_o_sha', 'attr_o',
                                                   'sinc_o', 'intel_o...
                               feature_types=None, gamma=None, gpu_id=None,
                               grow_policy=None, importance_type=None,
                               interaction_constraints=None, learning_rate=None,
                               max_bin=None, max_cat_threshold=None,
                               max_cat_to_onehot=None, max_delta_step=None,
                               max_depth=None, max_leaves=None,
                               min_child_weight=None, missing=nan,
                               monotone_constraints=None, n_estimators=100,
                               n_jobs=None, num_parallel_tree=None,
                               predictor=None, ...))])

In [ ]:
# fit the XGB model with train data.
XGB_pipeline = XGB_pipeline.fit(train_data, y)

[10:55:52] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.



In [ ]:
# display the score of the fitted XGB model.
XGB_pipeline.score(train_data, y)


1.0

### using Grid Search and XGBoost Classifier with Cross Validation:
I will use the Grid Search and XGBoost with Validation Set. I expect that it will give me the highest score, because the model try all possible values to know the optimal values, and fit the estimator (model) on your training set.

I'm going to specify some hyperparameters for the preprocessor, select features, and XGBoost classifier.


hyperparameters in the search space:
* strategy: "mean"
* n_estimaters: 30,40,50
* max_depth: 40,50,100

In [ ]:
# using grid search and XGB model.
# param_grid contain the search space that the grid search will find the best hyperparameters in.
param_grid = {
    'preprocessor__num__imputer__strategy': ['mean'],
    'my_classifier__n_estimators': [30,40,50],  
    'my_classifier__max_depth':[40,50,100] 
}

# apply the grid search model using XGB_pipeline and param_grid and auc_roc metric.
grid_search = GridSearchCV(XGB_pipeline, param_grid, cv=10, verbose=1, n_jobs=2, scoring='roc_auc')
# fitting the grid search with training data.
grid_search.fit(train_data, y)

print('best score {}'.format(grid_search.best_score_))
print('best score {}'.format(grid_search.best_params_))

Fitting 10 folds for each of 9 candidates, totalling 90 fits
[11:41:06] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

best score 0.8720482751383127
best score {'my_classifier__max_depth': 40, 'my_classifier__n_estimators': 50, 'preprocessor__num__imputer__strategy': 'mean'}


observation:

best hyperparameters after applying the grid search :

* strategy: "mean"
* n_estimaters: 50
* max_depth: 40

with accuracy in colab: 0.8720

,accuracy in kaggle public: 0.88278

and accuracy in kaggle private: 0.87674


### using Random Search and XGBoost Classifier with Cross Validation:
Random Search and XGBoost with Validation Set will be used. Because Random search works best for lower dimensional data and fits the estimator (model) on your training set, I expect it to give me the greatest score.

I'm going to specify some hyperparameters for the preprocessor, select features, and XGBoost classifier.

hyperparameters in the search space:
* strategy: "mean","most_frequency"
* n_estimaters: 20,30,40,50,60
* max_depth: 20,30,40,50,60


In [ ]:
#using random search to find the best hyperparamter combination in the XGB model.
param_rand = {
    'preprocessor__num__imputer__strategy': ['mean','most_frequent'],
    'my_classifier__n_estimators': [20,30, 40,50,60],  
    'my_classifier__max_depth':[20, 30,40,50,60] 
}
# apply the random search on the XGB_pieline and the param_rand and auc-roc metric.
rand_search = RandomizedSearchCV(XGB_pipeline, param_rand, cv=5, verbose=1, n_jobs=2, scoring='roc_auc')
# fit the model with training data.
rand_search.fit(train_data, y)

print('best score {}'.format(rand_search.best_score_))
print('best score {}'.format(rand_search.best_params_))

Fitting 5 folds for each of 10 candidates, totalling 50 fits
[12:07:39] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

best score 0.8720482751383127
best score {'my_classifier__max_depth': 40, 'my_classifier__n_estimators': 50, 'preprocessor__num__imputer__strategy': 'mean'}


observation:

best hyperparameters after applying the grid search :

* strategy: "mean"
* n_estimaters: 50
* max_depth: 40

with accuracy in colab: 0.8720

--------------------------------------------------------------------------------

### using bayes Search and XGBoost Classifier with Cross Validation:
I'll utilise Cross Validation with Bayesian Search and XGBoost. Because Bayesian Search discovers the extrema of objective functions that are expensive to evaluate and fits the estimator (model) on your training set, I expect it to give me the greatest score.

I'm going to specify some hyperparameters for the preprocessor, select features, and XGBoost classifier.

hyperparameter space:
* strategy: "mean","most_frequency"
* n_estimaters: [20-80]
* max_depth: [\20-60]

In [ ]:
# install the bayesSearch library.
!pip install scikit-optimize
from skopt import BayesSearchCV
from skopt.space import Real, Categorical, Integer


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100.3/100.3 KB 5.7 MB/s eta 0:00:00


In [ ]:
# make the BayesSearch pipeline that contain the XGB pipline and the paramters search space.
bayes_search = BayesSearchCV(
    XGB_pipeline,
    {
    'preprocessor__num__imputer__strategy': Categorical(['mean','most_frequent']),
    'my_classifier__n_estimators': Integer(20,80),  
    'my_classifier__max_depth':Integer(20,60) 
    },
    # number of trials 
    n_iter=3,
    random_state=0,
    verbose=1,
    # we still use 
    cv=10,
)
# fit the model with training data
bayes_search.fit(train_data, y)

print('best score {}'.format(bayes_search.best_score_))
print('best score {}'.format(bayes_search.best_params_))


Fitting 10 folds for each of 1 candidates, totalling 10 fits
[11:57:22] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[11:57:26] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[11:57:33] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[11:57:37] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[11:57:42] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[11:57:48] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[11:57:52] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[11:58:03] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[11:58:10] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[11:58:14] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

Fitting 10 folds for each of 1 candidat

observation:

best hyperparameters after applying the grid search :

* strategy: "most frequent"
* n_estimaters: 61
* max_depth: 41

with accuracy in colab: 0.8663

in kaggle public: 0.8791

kaggle private: 0.8699

-------------------------------------------------------------------------------

In [ ]:
# make a submission for the XGB using Bayesian search.
submission = pd.DataFrame()

submission['id'] = test_id

submission['match'] = rand_search.predict_proba(test_data)[:,1]

submission.to_csv('sample_submission_xgb_bayes1.csv', index=False)

### using bayes Search and XGBoost Classifier with Cross Validation with different hyperparameter space:
* strategy': ['mean']
* learning_rate' : [0.005, 0.001, 0.01, 0.02, 0.03, 0.05, 0.1, 0.2, 0.3]
* n_estimators' : [600,1000, 1100, 1500, 2000, 3000, 4000]
* nthread' : [1, 2, 3, 4, 5, 6, 7, 8, 9]
* subsample': [0.05, 0.2, 0.3, 0.6, 0.8, 0.9]
* olsample_bytree': [0.5, 0.6, 0.7, 0.8, 0.9, 1.0]




In [ ]:
# hyperparameter for XGBoost Classifier
# here we specify the search space 
# `__` denotes an attribute of the preceeding name
# (e.g. my_classifier__n_estimators means the `n_estimators` param for `my_classifier`)

bayes_search_2 = BayesSearchCV(
    XGB_pipeline,
    {
    'preprocessor__num__imputer__strategy': ['mean'],
    'my_classifier__learning_rate' : [0.005, 0.001, 0.01, 0.02, 0.03, 0.05, 0.1, 0.2, 0.3],
    'my_classifier__n_estimators' : [600,1000, 1100, 1500, 2000, 3000, 4000],
    'my_classifier__nthread' : [1, 2, 3, 4, 5, 6, 7, 8, 9],
    'my_classifier__subsample': [0.05, 0.2, 0.3, 0.6, 0.8, 0.9],
    'my_classifier__colsample_bytree': [0.5, 0.6, 0.7, 0.8, 0.9, 1.0],
},
    # number of trials 
    n_iter=3,
    random_state=0,
    verbose=1,
    # we still use 
    cv=10,
)

bayes_search_2.fit(train_data, y)

print('best score {}'.format(bayes_search_2.best_score_))
print('best score {}'.format(bayes_search_2.best_params_))

Fitting 10 folds for each of 1 candidates, totalling 10 fits
[15:14:37] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[15:16:06] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[15:17:28] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[15:18:56] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[15:20:23] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[15:21:53] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[15:23:14] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[15:24:37] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[15:26:00] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[15:27:26] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

Fitting 10 folds for each of 1 candidat

In [ ]:
# submission file for XGB model trail 2 with bayes search.
#best model accuracy on Kaggle
submission = pd.DataFrame()

submission['id'] = test_id

submission['match'] = bayes_search_2.predict_proba(test_data)[:,1]

submission.to_csv('sample_submission_xgb_bayes2.csv', index=False)

best hyperparameter:
* strategy': ['mean']
* learning_rate' : 0.05
* n_estimators' : 2000
* nthread' : 6
* subsample': 0.8
* colsample_bytree': 0.8

This model give the best accuracy on kaggle:

colab accuracy: 0.86579

public Score: 0.88296

Private score: 0.89129

### random forest
Random forest is a commonly-used machine learning algorithm trademarked by Leo Breiman and Adele Cutler, which combines the output of multiple decision trees to reach a single result. Its ease of use and flexibility have fueled its adoption, as it handles both classification and regression problems.

In [ ]:
# build a random forest pipeline.
RF_pipline = Pipeline(
    steps=[
        ('preprocessor', preprocessor),
        ('my_classifier', 
           RandomForestClassifier(),
        )
    ]
)
RF_pipline

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer()),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  ['gender', 'condtn', 'wave',
                                                   'round', 'position',
                                                   'positin1', 'order',
                                                   'partner', 'int_corr',
                                                   'samerace', 'age_o',
                                                   'race_o', 'pf_o_att',
                                                   'pf_o_sin', 'pf_o_int',
                                                   'pf_o_fun', 'pf_o_amb',
                                                   'pf_o_sha', 'attr_o',
                                                   'sinc_o', 'intel_o', 'fun_o',
                                                   'amb_o', 'shar_o', 'like_o',
                                                   'prob_o', 'met_o', 'age',
                                                   'field_cd', 'race', ...]),
                                                 ('cat',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='constant')),
                                                                  ('onehot',
                                                                   OneHotEncoder(handle_unknown='ignore'))]),
                                                  ['field', 'undergra', 'from',
                                                   'career'])])),
                ('my_classifier', RandomForestClassifier())])

### using Random Search and Random forest Classifier with Cross Validation:



Random Search and Random Forest with cross validation will be used. Because Random search works best for lower dimensional data and fits the estimator (model) on your training set, I expect it to give me the greatest score.

Hyperparamter space:
* n_estimators: 50,100,150,200
* max_dapth: 3,4,5,10,15,20
* criterion: 'gini','entropy'
* sample_split: 2,3,4,6,7,8

In [ ]:
# fit the random forest pipline with random search using the param_rand parameters space.

param_rand = {
    'my_classifier__n_estimators':[50,100,150,200],
    'my_classifier__max_depth':[3,4,5,10,15,20],
    'my_classifier__criterion':['gini','entropy'],
    'my_classifier__min_samples_split':[2,3,4,6,7,8],
}

rand_search_Rf = RandomizedSearchCV(XGB_pipeline, param_rand, cv=5, verbose=1, n_jobs=2, scoring='roc_auc')

rand_search_Rf.fit(train_data, y)

print('best score {}'.format(rand_search_Rf.best_score_))
print('best score {}'.format(rand_search_Rf.best_params_))

Fitting 5 folds for each of 10 candidates, totalling 50 fits
[12:30:52] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight", "criterion", "min_samples_split" } are not used.

best score 0.877002726594764
best score {'my_classifier__n_estimators': 150, 'my_classifier__min_samples_split': 4, 'my_classifier__max_depth': 10, 'my_classifier__criterion': 'gini'}


observation:

best hyperparameters after applying the grid search :

* n_estimators: 150
* max_dapth: 10
* criterion: 'gini'
* sample_split: 4

with accuracy in colab: 0.8770

in kaggle public: 0.87956

kaggle private: 0.88699

In [ ]:
# submission file using random forest with random search.
submission = pd.DataFrame()

submission['id'] = test_id

submission['match'] = rand_search_Rf.predict_proba(test_data)[:,1]

submission.to_csv('sample_submission_RF_rand2.csv', index=False)

## logistic regression
Logistic regression is used to handle the classification problems.

It is used in statistical software to understand the relationship between the dependent variable and one or more independent variables by estimating probabilities using a logistic regression equation.

It is often used for predictive analytics and modeling, and extends to applications in machine learning. Logistic regression is easier to implement, interpret, and very efficient to train.

In [ ]:
# param_log contains the hyperparameters sapce for the logistic regression model.
param_Log = {
    'preprocessor__num__imputer__strategy': ['mean'],
    'my_classifier__penalty' : [ 'l2','none'],
    'my_classifier__C' : [0.001,0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.1, 1],
    'my_classifier__solver' : ['newton-cg', 'saga','lbfgs'],
}
# build the logistic model pipeline.
log_pipline = Pipeline(
    steps=[
        ('preprocessor', preprocessor),
        ('my_classifier', 
           LogisticRegression(),
        )
    ]
)
log_pipline

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer()),
                                                                  ('scaler',
                                                                   StandardScaler())]),
                                                  ['gender', 'condtn', 'wave',
                                                   'round', 'position',
                                                   'positin1', 'order',
                                                   'partner', 'int_corr',
                                                   'samerace', 'age_o',
                                                   'race_o', 'pf_o_att',
                                                   'pf_o_sin', 'pf_o_int',
                                                   'pf_o_fun', 'pf_o_amb',
                                                   'pf_o_sha', 'attr_o',
                                                   'sinc_o', 'intel_o', 'fun_o',
                                                   'amb_o', 'shar_o', 'like_o',
                                                   'prob_o', 'met_o', 'age',
                                                   'field_cd', 'race', ...]),
                                                 ('cat',
                                                  Pipeline(steps=[('imputer',
                                                                   SimpleImputer(strategy='constant')),
                                                                  ('onehot',
                                                                   OneHotEncoder(handle_unknown='ignore'))]),
                                                  ['field', 'undergra', 'from',
                                                   'career'])])),
                ('my_classifier', LogisticRegression())])

### using Random Search and logistic Regression Classifier with Cross Validation:
 hyperparameter space:
* strategy: ['mean']
* penalty: [ 'l2','none']
* C: [0.001,0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.1, 1]
* solver: ['newton-cg', 'saga','lbfgs']

In [ ]:
## fit the logreg pipline with random search using the param_log parameters space.

rand_search_log = RandomizedSearchCV(log_pipline, param_Log, cv=10, verbose=1, n_jobs=2, scoring='roc_auc')

rand_search_log.fit(train_data, y)

print('best score {}'.format(rand_search_log.best_score_))
print('best score {}'.format(rand_search_log.best_params_))

Fitting 10 folds for each of 10 candidates, totalling 100 fits
best score 0.8639566006419729
best score {'preprocessor__num__imputer__strategy': 'mean', 'my_classifier__solver': 'lbfgs', 'my_classifier__penalty': 'l2', 'my_classifier__C': 0.1}


observation:

best hyperparameters after applying the random search :

* strategy: ['mean']
* penalty: [ 'l2']
* C: 0.1
* solver: ['lbfgs']
with accuracy in colab: 0.8639

in kaggle public: 0.87178

kaggle private: 0.85976

In [ ]:
#submission file for logistic regession.
submission = pd.DataFrame()

submission['id'] = test_id

submission['match'] = rand_search_log.predict_proba(test_data)[:,1]

submission.to_csv('sample_submission_log_1.csv', index=False)

###  XGB trail 2 with oversampling
The simplest oversampling method involves randomly duplicating examples from the minority class in the training dataset, referred to as Random Oversampling

In [ ]:
# libraries required for oversampling.
from imblearn.over_sampling import RandomOverSampler
from collections import Counter 
# take a copy from the training features and labels to apply oversampling on. 
X_train=train_data
y_train= y
oversample = RandomOverSampler(sampling_strategy='minority')
# fit the oversapmling on the copied train feature and label.
X_train , y_train = oversample.fit_resample(X_train, y_train)
# display the count of classes before and after the oversampling.
print('Original dataset shape:', Counter(y))
print('Resampple dataset shape:', Counter(y_train))

Original dataset shape: Counter({0: 4921, 1: 988})
Resampple dataset shape: Counter({0: 4921, 1: 4921})


### using bayes Search and XGB Classifier with Cross Validation and oversampling:
i expected that the model will have best accuracy with oversampling and 
Hyperparameter space:
* strategy: ['mean']
* learning_rate' : [0.005, 0.001, 0.01, 0.02, 0.03, 0.05, 0.1, 0.2, 0.3]
* n_estimators' : [600,1000, 1100, 1500, 2000, 3000, 4000]
* nthread' : [1, 2, 3, 4, 5, 6, 7, 8, 9]
* subsample': [0.05, 0.2, 0.3, 0.6, 0.8, 0.9]
* colsample_bytree': [0.5, 0.6, 0.7, 0.8, 0.9, 1.0]\


In [ ]:
# apply the XBG model pipeline after the oversampling using the bayes search. 
# hyperparameter for XGBoost Classifier
# here we specify the search space 
# `__` denotes an attribute of the preceeding name
# (e.g. my_classifier__n_estimators means the `n_estimators` param for `my_classifier`)

bayes_search_2 = BayesSearchCV(
    XGB_pipeline,
    {
    'preprocessor__num__imputer__strategy': ['mean'],
    'my_classifier__learning_rate' : [0.005, 0.001, 0.01, 0.02, 0.03, 0.05, 0.1, 0.2, 0.3],
    'my_classifier__n_estimators' : [600,1000, 1100, 1500, 2000, 3000, 4000],
    'my_classifier__nthread' : [1, 2, 3, 4, 5, 6, 7, 8, 9],
    'my_classifier__subsample': [0.05, 0.2, 0.3, 0.6, 0.8, 0.9],
    'my_classifier__colsample_bytree': [0.5, 0.6, 0.7, 0.8, 0.9, 1.0],
},
    # number of trials 
    n_iter=3,
    random_state=0,
    verbose=1,
    # we still use 
    cv=10,
)

bayes_search_2.fit(X_train, y_train)

print('best score {}'.format(bayes_search_2.best_score_))
print('best score {}'.format(bayes_search_2.best_params_))

Fitting 10 folds for each of 1 candidates, totalling 10 fits
[10:02:23] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[10:04:34] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[10:06:49] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[10:09:07] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[10:11:29] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[10:13:40] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[10:15:53] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[10:18:05] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[10:20:18] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

[10:22:27] WARNING: ../src/learner.cc:767: 
Parameters: { "class_weight" } are not used.

Fitting 10 folds for each of 1 candidat

In [ ]:
# submission file for XGB model trail 3 with bayes search and over sampling
submission = pd.DataFrame()

submission['id'] = test_id

submission['match'] = bayes_search_2.predict_proba(test_data)[:,1]

submission.to_csv('sample_submission_xgb_bayes3.csv', index=False)

observation:

best hyperparameters after applying the bayes search :

* strategy: ['mean']
* learning_rate' : [0.05
* n_estimators' : 2000
* nthread' : 6
* subsample': 0.8
* colsample_bytree': 0.8

with accuracy in colab: 0.9663

in kaggle public: 0.88357

kaggle private: 0.89005

it give me high accuracy on colab and in kaggle public score but the second highest accuracy on private kaggle score.

### using grid Search and XGB Classifier with Cross Validation and oversampling:
hyperparameter space:
* learning_rate': [0.1]
* max_depth': [8, 10, 12]
* n_estimators': [500]
* subsample': [0.9]
* reg_alpha': [0.1]
* reg_lambda': [0.1]
* gamma': [0.0]

same here the model expected to give high accuracy with over sampling and grid search in the hyperparameter space.

In [ ]:
# apply the XBG model pipeline after the oversampling using the grid search. 
full_pipline = Pipeline(
    steps = [
        ('preprocessor', preprocessor),
        ('my_classifier', XGBClassifier(objective='binary:logistic'))
    ]
)
full_pipline

parameters = {
    'my_classifier__learning_rate': [0.1],
    'my_classifier__max_depth': [8, 10, 12],
    'my_classifier__n_estimators': [500],
    'my_classifier__subsample': [0.9],
    'my_classifier__reg_alpha': [0.1],
    'my_classifier__reg_lambda': [0.1],
    'my_classifier__gamma': [0.0],
}

grid_search_XGB_new = GridSearchCV(full_pipline, parameters, cv=10, verbose=1, n_jobs=2, scoring='roc_auc')
grid_search_XGB_new.fit(X_train, y_train)

print('best score {}'.format(grid_search_XGB_new.best_score_))
print('best score {}'.format(grid_search_XGB_new.best_params_))


Fitting 10 folds for each of 3 candidates, totalling 30 fits
best score 0.9990772217281616
best score {'my_classifier__gamma': 0.0, 'my_classifier__learning_rate': 0.1, 'my_classifier__max_depth': 12, 'my_classifier__n_estimators': 500, 'my_classifier__reg_alpha': 0.1, 'my_classifier__reg_lambda': 0.1, 'my_classifier__subsample': 0.9}


In [ ]:
# submission file for XGB model trail 3 with random search and over sampling.
submission = pd.DataFrame()

submission['id'] = test_id

submission['match'] = rand_search.predict_proba(test_data)[:,1]

submission.to_csv('sample_submission_xgb_grid3.1.csv', index=False)

observation:

best hyperparameter:
* learning_rate': [0.1]
* max_depth': 12
* n_estimators': [500]
* subsample': [0.9]
* reg_alpha': [0.1]
* reg_lambda': [0.1]
* gamma': [0.0]

the model had the best accuracy on colab:0.9990

kaggle public score: 0.8844

kaggle private score: 0.88482


Thanks 😀